graphql/authorizations.ts

```diff
@@ -113,6 +155,75 @@ async function accessCheck(context: GraphQLContext, requiredRoles: Role[], model
     throw new ForbiddenError(`Requiring one of the following roles: ${requiredRoles}`);
 }
 
+// An Entity Role is a role that a user has on all edges from a specific entity (root)
+interface EntityRole {
+    role: Role;
+    hasRole(context: GraphQLContext, modelName: ResolverModelNames, root: any): Promise<boolean>;
+}
+
+// As we traverse multiple edges from the same entity, we might need to check the same entity roles for the same entity again,
+// as these checks can be costly involving a roundtrip to the db, we cache the evaluated role in the GraphQL context
+
+// undefined -> did not yet checked role
+// true -> user has role on entity
+// false -> user does not have role on entity
+type EvaluatedEntityRoles = { [role in Role]?: boolean };
+type RolesForEntities = Map<any /* root entity */, EvaluatedEntityRoles>;
+
+const EVALUATED_ROLES = Symbol();
```

Comment:
Out of curiosity, why is a `Symbol` used here? I read it's supposed to guarantee uniqueness, but why can't we just use e.g. `context.evaluatedRoles` to access the map instead? 🙂

Reply:
Cause I wanted to make sure that this attribute is only accessed and modified in authorizations.ts, as writing it is highly dangerous.